### PR TITLE
Finished voting system

### PR DESCRIPTION
**Pull request by Fir0w**
image

### CODE REVIEW

File: client/components/Post.jsx [hunk 1]
```diff
@@ -29,11 +42,12 @@ const Post = ({ setPostFormContent, userId, postId, profileName, postContent, up
...
     };
 
-    const getProfile = async () => {
+    const vote = async (voteOption) => {
 
         try {
-            const res = await axios.get(`/api/users/user?username=${profileName}`);
-            setProfile(res.data);
+            await axios.post(`/api/posts/vote`, { postId, userId: user.userId, voteOption });
+            const reponse = await axios.get(`/api/posts/vote?postId=${postId}`);
```
Comment (VicDosya) [inline client/components/Post.jsx:49]: response* :)

File: client/components/PostsList.jsx [hunk 2]
```diff
@@ -7,7 +7,7 @@ const PostsList = ({ setPostFormContent, postContent }) => {
 
 
         postContent.map((post) => {
-            return <Post key={post._id} setPostFormContent={setPostFormContent} userId={post.userId} postId={post._id} profileName={post.profileName} postContent={post.postContent} upVote={'up vote'} downVote={'down vote'} comment={`${post.commentsCount} comments`} timeStamp={new Date(post.createdAt).toLocaleString('en-GB', { hour: 'numeric', minute: 'numeric', second: 'numeric', year: 'numeric', month: 'numeric', day: 'numeric' })} />
+            return <Post key={post._id} setPostFormContent={setPostFormContent} userId={post.userId} postId={post._id} profileName={post.profileName} postContent={post.postContent} comment={`${post.commentsCount} comments`} timeStamp={new Date(post.createdAt).toLocaleString('en-GB', { hour: 'numeric', minute: 'numeric', second: 'numeric', year: 'numeric', month: 'numeric', day: 'numeric' })} />
```
Comment (VicDosya) [inline client/components/PostsList.jsx:10]: There's way too much props here,
can u make it better indentation ?
like this:
```
return <Post 
key={key}
bla={bla}
tick={tack}
>
```

File: server/controllers/voteController.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        if (!post)
+            return res.status(404).send({ message: "Post was not found" });
+
+        const voteExists = await Vote.findOne({ postId });
+
+        if (voteExists) {
+
+            if (voteOption === 'upvote') {
+
+                if (!await Vote.findOne({ postId, userId: { $elemMatch: { userId } } })) {
```
Comment (VicDosya) [inline server/controllers/voteController.js:24]: what does this mean ? the creator of post can't vote up his own posts ?
just asking out of curiosity

File: server/controllers/voteController.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+                    voteExists.userId.push({ userId, voteOption: 'upvote' });
+                    voteExists.votesCount++;
+                    await voteExists.save();
+                }
+
+                const test = voteExists.userId.map(e => e.userId);
```
Comment (VicDosya) [inline server/controllers/voteController.js:31]: what is this test ?

File: server/controllers/voteController.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                let result = 10;
+                for (let index = 0; index < voteExists.userId.length; index++) {
+                    const element = test[index];
+                    if (element === userId) {
+                        result = index
+                    }
+                }
+                if (voteExists.userId[result].voteOption === 'upvote') {
```
Comment (VicDosya) [inline server/controllers/voteController.js:64]: correct me if wrong,
if user presses up\down vote again (when he already voted), this should remove the vote ?

File: server/server.js [hunk 6]
```diff
@@ -19,6 +20,7 @@ app.use(express.urlencoded({ extended: false }));
 app.use(cookieParser());
 
 app.use('/api/posts', postRoutes);
+app.use('/api/posts/vote', voteRoues);
```
Comment (VicDosya) [inline server/server.js:23]: voteRoutes*

File: client/components/Comment.jsx [hunk 7]
```diff
@@ -1,14 +1,31 @@
-/* eslint-disable react/prop-types */
 import { useParams } from 'react-router-dom';
 import styles from './Post.module.css';
 import useAuth from './useAuthContext';
 import axios from 'axios';
+import { useCallback, useEffect, useState } from 'react';
```
Comment (VicDosya) [inline client/components/Comment.jsx:5]: explain to me the usage of useCallback in your application, why is it useful ?

**Review by VicDosya — COMMENTED** [review]
crazy progress. good job man, you're progressing lightning fast.
left a few comments for you ;)

File: server/controllers/voteController.js (hunk 3 above)
Comment (Fir0w) [reply server/controllers/voteController.js:24]: he can, if you talking about line 24 then i am checking if the user already exists in the collection if not i am adding him

File: server/controllers/voteController.js (hunk 4 above)
Comment (Fir0w) [reply server/controllers/voteController.js:31]: a bad naming of a variable changed it.

File: server/controllers/voteController.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            } else if (voteOption === 'downvote') {
+
+                if (!await Vote.findOne({ postId, userId: { $elemMatch: { userId } } })) {
+
+                    voteExists.userId.push({ userId, voteOption: 'downvote' });
+                    voteExists.votesCount--;
+                    await voteExists.save();
+                }
+
+                const test = voteExists.userId.map(e => e.userId);
```
[the remark this answers is not in the document]
Comment (Fir0w) [reply server/controllers/voteController.js:56]: a bad naming of a variable changed it.

File: server/controllers/voteController.js (hunk 5 above)
Comment (Fir0w) [reply server/controllers/voteController.js:64]: yes it is lowering the total by 1

File: client/components/Comment.jsx (hunk 7 above)
Comment (Fir0w) [reply client/components/Comment.jsx:5]: umm to be honest its to shut up ESLint from complaining without disabling him.

**Review by VicDosya — APPROVED** [review]

**Pull request merged**